lib/temporal/activity/context.rb

```diff
@@ -64,11 +94,54 @@ def name
 
       private
 
-      attr_reader :connection, :metadata
+      attr_reader :connection, :metadata, :start_to_close_deadline, :heartbeat_thread_pool, :config, :heartbeat_mutex, :last_heartbeat_details
 
       def task_token
         metadata.task_token
       end
+
+      def heartbeat_throttle_interval
+        interval = if metadata.heartbeat_timeout > 0
+          metadata.heartbeat_timeout * 0.8
+        else
+          config.timeouts[:default_heartbeat_throttle_interval]
+        end
+
+        [interval, config.timeouts[:max_heartbeat_throttle_interval]].min
+      end
+
+      def send_heartbeat(details)
+        begin
+          response = connection.record_activity_task_heartbeat(
+            namespace: metadata.namespace,
+            task_token: task_token,
+            details: details)
+          if response.cancel_requested
+            logger.info('Activity has been canceled', metadata.to_h)
+            @cancel_requested = true
+          end
+        rescue => error
+          Temporal::ErrorHandler.handle(error, config, metadata: metadata)
+          raise
+        end
+      end
+
+      def schedule_heartbeat(delay)
+        heartbeat_thread_pool.schedule([metadata.workflow_run_id, metadata.id, metadata.attempt], delay) do
+          details = heartbeat_mutex.synchronize do
+            @heartbeat_scheduled = nil
```

Comment:
Suppose a user is sending a heartbeat every 10ms and the heartbeat throttle interval is 4 seconds.
If I'm reading this right, we'll send heartbeats at 10ms, 4.010s, 4.020s, 8.020s, 8.030s, ... because we clear the scheduled heartbeat-- so the next time the user heartbeats, we'll  immediately send it.
Is there any reason not to schedule the next one immediately so that we end up sending half as many heartbeats?

In either case, probably adding a high-level example will help folks understand this code.

Reply:
I'll add an example of this and confirm the behavior. We do schedule the next one immediately because when the timer wakes up, it calls the function recursively (from the callback so it's not typical recursion).

Your example is tricky to explain precisely because there's a race condition between the scheduled heartbeat and the next one coming in. But here's roughly how it would work:

10ms: first heartbeat request
- gets sent immediately
- assuming that call takes 100ms to complete a timer will be scheduled to fire at 4110ms to send any future heartbeat

20ms: second heartbeat request
- value saved for later

... same thing happens again, last heartbeat value gets updated over and over

4100ms: 410th heartbeat comes in, value saved
4110ms: timer goes off (let's assume it wins the race condition against the 411th heartbeat request)
- since there is a saved heartbeat, we send it immediately, this is the 410th heartbeat's value
- assume sending the heartbeat takes 100ms again
- timer scheduled for 8210ms (4110 + 100 + 4000)

Also at 4110ms on another thread: this 411th heartbeat is blocked obtaining the mutex since a heartbeat is being sent  on the first thread until 4210ms (when the heartbeat finishes being sent)

4210ms: 411th heartbeat finally gets the mutex. Because another timer has been scheduled, its value saved and not sent immediately 

... same thing happens again

The scheduling chain would only be broken if there's no saved heartbeat when the timer wakes up (i.e., no heartbeat was sent during the entire throttling interval). In that case, it has no heartbeat to send, so therefore does not send it and does not recursively schedule a future heartbeat.

Reply:
Added a simpler, easier to follow example into a comment